### PR TITLE
[libc][math][c23] Add exp2m1f16 C23 math function

### PR DESCRIPTION
**Pull request by overmighty**
Part of #95250.

### CODE REVIEW

**Review by jhuber6 — COMMENTED** [review]
Could you go ahead and add this to the GPU entrypoints? I'm trying to keep the mostly in sync. (Though I need to figure out why a few of them fail).

**Comment by overmighty** [comment]
Done. I haven't run the GPU build in a while so I haven't tested it yet, as it'll probably take some time to run the GPU build again.

**Comment by jhuber6** [comment]
> Done. I haven't run the GPU build in a while so I haven't tested it yet, as it'll probably take some time to run the GPU build again.

I'll test it in a bit.

File: libc/config/gpu/entrypoints.txt [hunk 1]
```diff
@@ -486,6 +486,7 @@ if(LIBC_TYPES_HAS_FLOAT16)
     libc.src.math.ceilf16
     libc.src.math.copysignf16
     libc.src.math.exp10f16
+    libc.src.math.exp10m1f16
```
Comment (jhuber6) [inline libc/config/gpu/entrypoints.txt:489]: Is this the right name?

Comment (overmighty) [reply libc/config/gpu/entrypoints.txt:489]: Oops.

Comment (overmighty) [reply libc/config/gpu/entrypoints.txt]: Fixed.

**Comment by jhuber6** [comment]
```[2220/2586] Running hermetic test libc.test.src.math.smoke.exp2m1f16_test.__hermetic__
FAILED: libc/test/src/math/smoke/libc.test.src.math.smoke.exp2m1f16_test.__hermetic__-cmd /home/jhuber/Documents/llvm/llvm-project/build/runtimes/runtimes-amdgcn-amd-amdhsa-bins/libc/test/src/math/smoke/libc.test.src.math.smoke.exp2m1f16_test.__hermetic__-cmd 
cd /home/jhuber/Documents/llvm/llvm-project/build/runtimes/runtimes-amdgcn-amd-amdhsa-bins/libc/test/src/math/smoke && /home/jhuber/Documents/llvm/llvm-project/build/bin/amdhsa-loader /home/jhuber/Documents/llvm/llvm-project/build/runtimes/runtimes-amdgcn-amd-amdhsa-bins/libc/test/src/math/smoke/libc.test.src.math.smoke.exp2m1f16_test.__hermetic__.__build__
[==========] Running 3 tests from 1 test suite.
[ RUN      ] LlvmLibcExp2m1f16Test.SpecialNumbers
/home/jhuber/Documents/llvm/llvm-project/libc/test/src/math/smoke/exp2m1f16_test.cpp:38: FAILURE
Failed to match (__llvm_libc_20_0_0_git::exp2m1f16(neg_zero)) against LIBC_NAMESPACE::testing::getMatcher< LIBC_NAMESPACE::testing::TestCond::EQ>((neg_zero)).
Expected floating point value: 0x8000 = (S: 1, E: 0x0000, M: 0x0000)
Actual floating point value: 0x0000 = (S: 0, E: 0x0000, M: 0x0000)
```
Hm, works fine on NVPTX but fails on AMDGPU. Wonder why the sign is off.

**Comment by jhuber6** [comment]
I poked at it but couldn't figure out why it's giving `0.0` instead of `-0.0`. Seems to work on NVPTX so I don't know what's different here. Ruled out `__builtin_rint` and optimizations, so it might just be some weird lowering. @arsenm might have a guess? Otherwise, feel free to remove (or comment it out with a note) for the GPU case so you're not blocked on that.

**Comment by arsenm** [comment]
> I poked at it but couldn't figure out why it's giving `0.0` instead of `-0.0`. Seems to work on NVPTX so I don't know what's different here. Ruled out `__builtin_rint` and optimizations, so it might just be some weird lowering. @arsenm might have a guess? Otherwise, feel free to remove (or comment it out with a note) for the GPU case so you're not blocked on that.

exp2 should be correctly lowered and it's just a bug if it's not

**Comment by overmighty** [comment]
This doesn't use `@llvm.exp2.f16`. We have our own implementations of exponential functions using polynomial approximations. I'll try to post IR and AMDGPU assembly output tomorrow if Joseph doesn't beat me to it.

File: libc/spec/stdc.td [hunk 2]
```diff
@@ -596,6 +596,7 @@ def StdC : StandardSpec<"stdc"> {
           GuardedFunctionSpec<"exp2f16", RetValSpec<Float16Type>, [ArgSpec<Float16Type>], "LIBC_TYPES_HAS_FLOAT16">,
 
           FunctionSpec<"exp2m1f", RetValSpec<FloatType>, [ArgSpec<FloatType>]>,
+          GuardedFunctionSpec<"exp2m1f16", RetValSpec<Float16Type>, [ArgSpec<Float16Type>], "LIBC_TYPES_HAS_FLOAT16">,
```
Comment (lntue) [inline libc/spec/stdc.td:599]: also update new header gen.

**Comment by overmighty** [comment]
> exp2 should be correctly lowered and it's just a bug if it's not. Is the bug in whatever this emitted or the llvm.exp2.f16 lowering? Can you post IR/ISA output?

Sorry for the (long) delay, here's the IR output: https://gist.github.com/overmighty/20ec6b34f36d6b02c8009d85fa6e0794.

**Comment by arsenm** [comment]
> > exp2 should be correctly lowered and it's just a bug if it's not. Is the bug in whatever this emitted or the llvm.exp2.f16 lowering? Can you post IR/ISA output?
> 
> 
> 
> Sorry for the (long) delay, here's the IR output: https://gist.github.com/overmighty/20ec6b34f36d6b02c8009d85fa6e0794.

Why are there any volatile accesses? There shouldn't be any z

But still, this should just emit llvm.exp2

**Comment by overmighty** [comment]
> Why are there any volatile accesses? There shouldn't be any z

Some of them are from https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/rounding_mode.h#L67. I'm not sure about the others.

> But still, this should just emit llvm.exp2

I don't think any higher math function implementation in LLVM libc emits the respective LLVM intrinsic currently.

**Comment by arsenm** [comment]
> > Why are there any volatile accesses? There shouldn't be any z
> 
> Some of them are from https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/rounding_mode.h#L67. I'm not sure about the others.
> 

These volatiles should be removed. The tests will work just fine without the forced stack access

**Comment by jhuber6** [comment]
> > > Why are there any volatile accesses? There shouldn't be any z
> > 
> > 
> > Some of them are from https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/rounding_mode.h#L67. I'm not sure about the others.
> 
> These volatiles should be removed. The tests will work just fine without the forced stack access

The usage seems to be with the rounding mode check, so it's important that nothing gets constant propagated. I actually don't know how the rounding mode i handled when 

> > > Why are there any volatile accesses? There shouldn't be any z
> > 
> > 
> > Some of them are from https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/rounding_mode.h#L67. I'm not sure about the others.
> 
> These volatiles should be removed. The tests will work just fine without the forced stack access

These try to get the rounding mode through math operations so it's really important that it's not constant propagated. If you make it a constant then it will go through the LLVM optimizer which just assumes it's to-nearest, see https://godbolt.org/z/ehejrv1zK. We could probably avoid the volatile with some inline assembly or something.

**Comment by arsenm** [comment]
> The usage seems to be with the rounding mode check, so it's important that nothing gets constant propagated. I actually don't know how the rounding mode i handled when

They won't be if you are using strictfp (as you must be if you depend on the rounding mode). If you do know the rounding mode, they will correctly fold.


> These try to get the rounding mode through math operations so it's really important that it's not constant propagated. If you make it a constant then it will go through the LLVM optimizer which just assumes it's to-nearest, see https://godbolt.org/z/ehejrv1zK. We could probably avoid the volatile with some inline assembly or something.

You need to enable fenv access

**Comment by jhuber6** [comment]
> You need to enable fenv access

That doesn't work on the GPU targets yet https://godbolt.org/z/6n3svs1ne. But I guess NVPTX doesn't have `fenv` access and I never got around to merging the AMDGPU version.

**Comment by arsenm** [comment]
> That doesn't work on the GPU targets yet https://godbolt.org/z/6n3svs1ne. But I guess NVPTX doesn't have `fenv` access and I never got around to merging the AMDGPU version.

Correct, but you shouldn't design around working around unimplemented features. Full GPU strictfp support isn't coming anytime soon

**Comment by arsenm** [comment]
> > That doesn't work on the GPU targets yet https://godbolt.org/z/6n3svs1ne. But I guess NVPTX doesn't have `fenv` access and I never got around to merging the AMDGPU version.
> 
> Correct, but you shouldn't design around working around unimplemented features. Full GPU strictfp support isn't coming anytime soon. It is still wrong to rely on this check without it

The other really broken thing is the -fexperimental-strict-floating-point flag doesn't do anything other than for the hardcoded targets in clang, making it useless as a bringup aid

**Comment by lntue** [comment]
> > > That doesn't work on the GPU targets yet https://godbolt.org/z/6n3svs1ne. But I guess NVPTX doesn't have `fenv` access and I never got around to merging the AMDGPU version.
> > 
> > 
> > Correct, but you shouldn't design around working around unimplemented features. Full GPU strictfp support isn't coming anytime soon. It is still wrong to rely on this check without it
> 
> The other really broken thing is the -fexperimental-strict-floating-point flag doesn't do anything other than for the hardcoded targets in clang, making it useless as a bringup aid

We can omit the gpu targets for now.  I'm going to add `LIBC_MATH_DEFAULT_ROUNDING_ONLY` option to cleanly omit all the directional rounding codes for targets requesting it.

**Comment by overmighty** [comment]
Rebased and added the last 3 commits.

**Review by lntue — APPROVED** [review]

**Pull request merged**